File: libs/go/pkg/autonomy/truth.go
package autonomy

import (
	"errors"
	"strings"
)

// TruthEngine is the Phase 08 verification contract.
//
//	true, nil   = verification passed
//	false, nil  = verification failed (known false / denied)
//	false, err  = verification unavailable / unknown / malformed
//
// UNKNOWN ≠ TRUE. Callers MUST NOT treat (false, error) as verified.
type TruthEngine interface {
	Verify(text string) (bool, error)
}

// DevelopmentTruth is a deterministic DEVELOPMENT / TEST rule engine.
// It is not internet verification, source corroboration, or LLM fact-checking.
type DevelopmentTruth struct{}

const TruthSafeFixture = "DEV_TRUTH_FIXTURE: local observation"

var (
	ErrTruthUnavailable = errors.New("TRUTH_UNAVAILABLE")
	truthDenied         = []string{"the earth is flat", "2+2=5", "known-false:"}
)

func (DevelopmentTruth) Kind() string { return "DEVELOPMENT_RULE_ENGINE" }

func (DevelopmentTruth) Verify(text string) (bool, error) {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return false, ErrTruthUnavailable
	}
	lower := strings.ToLower(trimmed)
	for _, d := range truthDenied {
		if strings.Contains(lower, d) {
			return false, nil
		}
	}
	if trimmed == TruthSafeFixture {
		return true, nil
	}
	return false, ErrTruthUnavailable
}

File: apps/web/lib/autonomy-control/truth.ts
/** DEVELOPMENT / TEST Truth engine. Not internet fact-checking. UNKNOWN ≠ TRUE. */

export const TRUTH_SAFE_FIXTURE = "DEV_TRUTH_FIXTURE: local observation";

const DENIED = ["the earth is flat", "2+2=5", "known-false:"];

export class TruthUnavailable extends Error {
  readonly code = "TRUTH_UNAVAILABLE";
  constructor(message: string) {
    super(message);
    this.name = "TruthUnavailable";
  }
}

export type TruthResult = { ok: boolean; error?: TruthUnavailable };

export class DevTruthEngine {
  readonly kind = "DEVELOPMENT_RULE_ENGINE";

  verify(text: string): TruthResult {
    const trimmed = (text ?? "").trim();
    if (!trimmed) return { ok: false, error: new TruthUnavailable("empty input cannot be verified") };
    const lower = trimmed.toLowerCase();
    if (DENIED.some((d) => lower.includes(d))) return { ok: false };
    if (trimmed === TRUTH_SAFE_FIXTURE) return { ok: true };
    return { ok: false, error: new TruthUnavailable("claim is unknown to the development rule engine") };
  }
}
